site/_posts/2020-07-22-securely-implement-oauth-vuejs.md

```diff
@@ -0,0 +1,1213 @@
+</div>
+```
+
+Here is how it our application looks like now, we can now login and logout with a click. Here's the application when you are signed out:
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/when-logged-out.png" alt="application when logged out." class="img-fluid" figure=false %}
+
+And here's the application when you are signed in (if you signed up with `richard@fusionauth.io`):
+{% include _image.liquid src="/assets/img/blogs/oauth-vuejs/when-logged-in.png" alt="The application when logged in." class="img-fluid" figure=false %}
+
+## Changing user info
+
+This last section deals with setting user data from our application.
+
+We will create the `/set-user-data` route; inside `routes` create `set-user-data.js` and add this code to it:
+
+```javascript
+const express = require("express");
+const router = express.Router();
+const axios = require("axios");
+const qs = require("query-string");
+router.post("/", (req, res) => {
+  // POST request to /introspect endpoint
+  axios
+    .post(
+      `http://localhost:${process.env.FUSIONAUTH_PORT}/oauth2/introspect`,
+      qs.stringify({
+        client_id: process.env.CLIENT_ID,
+        token: req.session.token,
+      })
+    )
+    .then((response) => {
+      let introspectResponse = response.data;
+    
+      // PATCH request to /registration endpoint
+      axios.patch(
+        `http://localhost:${process.env.FUSIONAUTH_PORT}/api/user/registration/${introspectResponse.sub}/${process.env.APLICATION_ID}`,
+        {
+          registration: {
+            data: req.body,
+          },
+        },
+        {
+          headers: {
+            Authorization: process.env.API_KEY,
+          },
+        }
+      ).catch(err=>{
+          console.log(err)
+      })
+    })
+    .catch((err) => {
+      console.error(err);
+    });
+
+});
+module.exports = router;
+```
+
+To ensure that we are updating the user currently logged in, we extract the token from our server which is done by making a **POST** request to the `oauth/introspect` endpoint; this is similar to what we did in the `user` route. 
+
+Once this request is successful, we make a **PATCH** request to `/api/user/registration` API. If you go through the [User Registration docs](https://fusionauth.io/docs/v1/tech/apis/registrations#update-a-user-registration), you will find that this API accepts both **PUT** and **PATCH** requests but here we are using **PATCH** since we only want to update a single part of the resource and **PATCH** will merge the request parameters into the existing object. 
+
+The `data` to send is stored inside the `registration` object which takes it value from `req.body`. This `registration` represents a user's association with an application. The `data` attribute allows us to store arbitrary data related to a user's registration in an application. We will be sending user data from our Vue app to the Express server via JSON in the body of a **POST** HTTP message.
```

Comment:
"takes it value" should probably be "takes its value". 

Also, this says **POST** but aren't we doing **PATCH**?

Reply:
The Vue app is `POST`ing to the express app, which is then `PATCH`ing to the FusionAuth app.

Will rework to be less confusing.